pkg/ring/ring.go

```diff
@@ -363,29 +363,74 @@ func (r *Ring) GetReplicationSetForOperation(op Operation) (ReplicationSet, erro
 		return ReplicationSet{}, ErrEmptyRing
 	}
 
-	// Calculate the number of required ingesters;
-	// ensure we always require at least RF-1 when RF=3.
-	numRequired := len(r.ringDesc.Ingesters)
-	if numRequired < r.cfg.ReplicationFactor {
-		numRequired = r.cfg.ReplicationFactor
-	}
-	maxUnavailable := r.cfg.ReplicationFactor / 2
-	numRequired -= maxUnavailable
-
-	ingesters := make([]IngesterDesc, 0, len(r.ringDesc.Ingesters))
+	// Build the initial replication set, excluding unhealthy instances.
+	healthyInstances := make([]IngesterDesc, 0, len(r.ringDesc.Ingesters))
+	zoneFailures := make(map[string]struct{})
 	for _, ingester := range r.ringDesc.Ingesters {
 		if r.IsHealthy(&ingester, op) {
-			ingesters = append(ingesters, ingester)
+			healthyInstances = append(healthyInstances, ingester)
+		} else {
+			zoneFailures[ingester.Zone] = struct{}{}
 		}
 	}
 
-	if len(ingesters) < numRequired {
-		return ReplicationSet{}, ErrTooManyFailedIngesters
+	// Max errors and max unavailable zones are mutually exclusive. We initialise both
+	// to 0 and then we update them whether zone-awareness is enabled or not.
+	maxErrors := 0
+	maxUnavailableZones := 0
+
+	if r.cfg.ZoneAwarenessEnabled {
+		// Given data is replicated to RF different zones, we can tolerate a number of
+		// RF/2 failing zones. However, we need to protect from the case the ring currently
+		// contains instances in a number of zones < RF.
+		numReplicatedZones := util.Min(len(r.ringZones), r.cfg.ReplicationFactor)
+		minSuccessZones := (numReplicatedZones / 2) + 1
+		maxUnavailableZones = minSuccessZones - 1
+
+		if len(zoneFailures) > maxUnavailableZones {
+			return ReplicationSet{}, ErrTooManyFailedIngesters
+		}
+
+		if len(zoneFailures) > 0 {
+			// We remove all instances (even healthy ones) from zones with at least
+			// 1 failing ingester. Due to how replication works when zone-awareness is
+			// enabled (data is replicated to RF different zones), there's no benefit in
+			// querying healthy instances from "failing zones". A zone is considered
+			// failed if there is single error.
```

Comment:
I don't follow the logic behind this.
If I have 1 failed replica in zone A, and 1 failed in zone B, and none failed in zone C, won't this cause an outage for some timeseries when there would have been none in some cases(i.e. the instance in A and the instance in B didn't host a replica of any of the same series). With this logic of removing everything from a zone if there is a failure, won't this make it more likely to have an outage in some cases?

Reply:
It depends.

Assuming RF=3, if `-distributor.shard-by-all-labels` is disabled, then this function is not used and we only query the 3 ingesters holding all the series for the queried metric (and we just need 2 success out of 3 queried ingesters).

However, if `-distributor.shard-by-all-labels` is enabled, we don't know which ingesters hold all the series for a given metric name, so we need to query all of them. When zone-awareness is disabled, we need to query all ingester minus 1, while when zone-awareness is enabled we need to query all ingesters in all zones minus 1 (thus tolerating all ingesters in 1 zone failing).

About shuffle sharding, this function is called after the subring has been selected.

Am I missing anything?

Reply:
Okay, I follow now, and am convinced the logic is correct, but maybe less enthusiastic about some of the implications, but since the value of  _all ingesters_ when -distributor.shard-by-all-labels is enabled is the subring seems more reasonable.